### ISSUE TITLE
Incorrect view context means ERB blocks cannot be passed into presenters

### ISSUE THREAD

**Issue by mattbrictson** [endofunky/oprah#1]
Great gem! I'm wondering if you have run into a problem I had when trying to build something similar. Basically, passing a block from an ERB to a presenter produces incorrect output. (See detailed example below.)

This is because the `view_context` that is given to `Oprah::Presenter` by the controller is different from the `view_context` used by the ERB view. Every time a view is rendered by Rails, it instantiates a new view context.

The issue is summarized in this Draper gem thread: https://github.com/drapergem/draper/issues/124. Their solution was a monkey patch to override the controller's `view_context` method to store the new context in a thread-local variable. The presenter objects then use the thread-local to always have access to the latest view context. Code here: https://github.com/drapergem/draper/commit/0b03d9c21ed3dd384e3e48adeae6463ee4d12aa4

-----

Steps to reproduce

Consider this contrived example of a presenter method that accepts a block:

```ruby
class PostPresenter < Oprah::Presenter
  def paragraph(&block)
    h.content_tag(:p, &block)
  end
end
```

And its use in a view:

```erb
<%= @post.paragraph do %>
  Hello!
<% end %>
```

The rendered output is wrong. The contents of the block are rendered twice, resulting in this output:

```
  Hello!
<p>
  Hello!
</p>
```

If you put the same code in a helper instead, it works fine:

```ruby
module PostsHelper
  def paragraph(&block)
    content_tag(:p, &block)
  end
end
```

```erb
<%= paragraph do %>
  Hello!
<% end %>
```

Result:

```
<p>
  Hello!
</p>
```

One workaround is to instantiate the presenter inside the view rather than inside the controller, in order to get the correct view context, but this feels like a huge hack:

```erb
<% @post = PostPresenter.new(@post, view_context: self) %>
```

Or, perhaps the monkey patch solution is not too bad, since the [Rails documentation itself](http://api.rubyonrails.org/v4.2.7.1/classes/ActionView/Rendering.html#method-i-view_context) encourages developers to override the `view_context` method:

> Override this method in a module to change the default behavior.

**Comment by endofunky** [comment]
Interesting. This is something we haven't done so far so I didn't stumble upon this yet. I'm trying to get a failing test to work right now and will look into possible options then. It'd be nice to find a different solution to draper's, who are essentially keeping a global variable.

**Comment by endofunky** [comment]
So it seems one work-around would be to keep track of the latest view context in the controller instead of a thread-local variable on the `Oprah` module, which is what draper seems to be doing.

We're already using `#oprah_view_context` instead of `#view_context` directly, so the `ControllerHelper` could pass a proxy into the presenter instead.

Example:

```ruby
class PostPresenter < Oprah::Presenter
  def paragraph(&block)
    h.content_tag(:p, &block)
  end
end

class ViewContextProxy < ActiveSupport::ProxyObject
  def initialize(controller)
    @controller = controller
  end

  def method_missing(meth, *args, &block)
    @controller.oprah_view_context.__send__(meth, *args, &block)
  end
end

class HomeController < ApplicationController
  def index
    @post = PostPresenter.new(nil, view_context: ViewContextProxy.new(self))
  end

  def view_context
    @oprah_view_context = super
  end

  def oprah_view_context
    @oprah_view_context || view_context
  end
end
```

**Comment by endofunky** [comment]
There is now a fix on master (76a8bc0e448406b5862c6c0cb9acfddcba57f951) for this using the proxy solution from above, avoiding thread-local, global variables.

I've included a dummy Rails app to `test/` and it renders your example correctly now.

Will push this to RubyGems as part of v0.1.3.

**Comment by mattbrictson** [comment]
Works well! Thanks 😀
